### INTRODUCTION

#### Predicting the mushrooms mainly through logistic regression
#### Mushroom class was classified into two categories  poisonous(0) and edible(1)
#### Steps taken in preprocessing includes Data cleaning,etc
#### All our variables in this dataset are categorical

### SIDE NOTE
#### You can leave your question about any unclear part in the comment section
#### Any correction will be highly welcomed

In [77]:
#Importing the necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.metrics import confusion_matrix

### LOADING THE DATASET

In [78]:
path = 'C:\\Users\ABOLARIN BUKOLA\Documents\Data set\ml\mushrooms.csv'

df = pd.read_csv(path)

In [79]:
df.head()

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0     p         x           s         n       t    p               f   
1     e         x           s         y       t    a               f   
2     e         b           s         w       t    l               f   
3     p         x           y         w       t    p               f   
4     e         x           s         g       f    n               f   

  gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0            c         n          k  ...                        s   
1            c         b          k  ...                        s   
2            c         b          n  ...                        s   
3            c         n          n  ...                        s   
4            w         b          k  ...                        s   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      w                      w         p          w   
1                      w                      w         p          w   
2                      w                      w         p          w   
3                      w                      w         p          w   
4                      w                      w         p          w   

  ring-number ring-type spore-print-color population habitat  
0           o         p                 k          s       u  
1           o         p                 n          n       g  
2           o         p                 n          n       m  
3           o         p                 k          s       u  
4           o         e                 n          a       g  

[5 rows x 23 columns]

In [80]:
df.describe()

class cap-shape cap-surface cap-color bruises  odor gill-attachment  \
count   8124      8124        8124      8124    8124  8124            8124   
unique     2         6           4        10       2     9               2   
top        e         x           y         n       f     n               f   
freq    4208      3656        3244      2284    4748  3528            7914   

       gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
count          8124      8124       8124  ...                     8124   
unique            2         2         12  ...                        4   
top               c         b          b  ...                        s   
freq           6812      5612       1728  ...                     4936   

       stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
count                    8124                   8124      8124       8124   
unique                      9                      9         1          4   
top                         w                      w         p          w   
freq                     4464                   4384      8124       7924   

       ring-number ring-type spore-print-color population habitat  
count         8124      8124              8124       8124    8124  
unique           3         5                 9          6       7  
top              o         p                 w          v       d  
freq          7488      3968              2388       4040    3148  

[4 rows x 23 columns]

### DEALING WITH MISSING VALUES

In [81]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8124 entries, 0 to 8123
Data columns (total 23 columns):
class                       8124 non-null object
cap-shape                   8124 non-null object
cap-surface                 8124 non-null object
cap-color                   8124 non-null object
bruises                     8124 non-null object
odor                        8124 non-null object
gill-attachment             8124 non-null object
gill-spacing                8124 non-null object
gill-size                   8124 non-null object
gill-color                  8124 non-null object
stalk-shape                 8124 non-null object
stalk-root                  8124 non-null object
stalk-surface-above-ring    8124 non-null object
stalk-surface-below-ring    8124 non-null object
stalk-color-above-ring      8124 non-null object
stalk-color-below-ring      8124 non-null object
veil-type                   8124 non-null object
veil-color                  8124 non-null object
ring-number

#### This dataset is clean it does not have any missing value

### DUMMY VARIABLES

#### Our target column 'class' contains two unique values 'e' and 'p' which we will map to '1' and '0' respectively
#### note 'e' stands for edible while 'p' for  poisonous

In [82]:
#Replacing e with 1
#Replacing  p with 0
df['class'] = df['class'].map({'e':1, 'p':0})

In [83]:
#Note that all our independent varibles are categorical variables
#Using attribute get_dummies to convert conver our category variables into dummy indicator
df_dummies = pd.get_dummies(df, drop_first = True)

In [84]:
df_dummies.head()

class  cap-shape_c  cap-shape_f  cap-shape_k  cap-shape_s  cap-shape_x  \
0      0            0            0            0            0            1   
1      1            0            0            0            0            1   
2      1            0            0            0            0            0   
3      0            0            0            0            0            1   
4      1            0            0            0            0            1   

   cap-surface_g  cap-surface_s  cap-surface_y  cap-color_c  ...  \
0              0              1              0            0  ...   
1              0              1              0            0  ...   
2              0              1              0            0  ...   
3              0              0              1            0  ...   
4              0              1              0            0  ...   

   population_n  population_s  population_v  population_y  habitat_g  \
0             0             1             0             0          0   
1             1             0             0             0          1   
2             1             0             0             0          0   
3             0             1             0             0          0   
4             0             0             0             0          1   

   habitat_l  habitat_m  habitat_p  habitat_u  habitat_w  
0          0          0          0          1          0  
1          0          0          0          0          0  
2          0          1          0          0          0  
3          0          0          0          1          0  
4          0          0          0          0          0  

[5 rows x 96 columns]

### LOGISTIC REGRESSION
#### Here we will create, fit and train our model in addition to that we will try to predict using the already trained model

In [85]:
#Declaring our target variable as y
#Declaring our independent variables as x
x = df_dummies.drop('class', axis = 1)
y = df_dummies['class']

In [86]:
#Selecting the model
reg = LogisticRegression()

In [87]:
#Splitting our dataset into train and test datasets 
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size = 0.2, random_state = 24)

In [88]:
#training the model with x_train and y_train
reg.fit(x_train, y_train)

C:\Users\ABOLARIN BUKOLA\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [89]:
#Predicting with our already trained model using x_test
y_hat = reg.predict(x_test)

In [90]:
#Getting the accuracy of our model
acc = metrics.accuracy_score(y_hat, y_test)
acc

0.9993846153846154

In [91]:
#The intercept for our regression
reg.intercept_

array([0.0174488])

In [92]:
#Coefficient for all our variables
reg.coef_

array([[-0.66281319,  0.33158298,  0.14557179,  0.60383078,  0.38661603,
        -0.97163288, -0.57330048, -0.48527047,  1.17209111, -0.1329527 ,
         0.13636038,  0.39346571, -0.62091301,  0.55368717,  0.46655004,
        -0.19444546,  0.56754541,  0.45811434, -3.44965213, -2.97976032,
         1.58340058, -0.26507977,  4.07537657, -3.05687681, -1.3320444 ,
        -1.31933919, -0.34057976,  2.88370506, -3.34983184,  0.63792642,
         0.13005592, -0.02904466,  0.48786068,  0.85085914,  0.19464443,
         0.68512579, -0.67247058,  0.28199285,  0.47272519, -0.11877884,
         1.00644384, -2.28704333,  1.51261178, -0.79321797,  1.359912  ,
        -1.82191154,  0.7474001 , -0.26792871, -0.66443736, -0.12296563,
        -1.4688839 , -0.26507977,  0.53300638,  0.34158523,  0.26125025,
         0.5301413 , -0.44454875,  0.09332687, -0.81993956, -0.26507977,
         0.43646321,  0.28597076,  1.12867882,  0.5301413 , -0.60011497,
        -0.2863821 , -1.07778918,  0.22879573,  0.3

### CONFUSION MATRIX

In [93]:
cm = confusion_matrix(y_hat, y_test)
cm

array([[755,   0],
       [  1, 869]], dtype=int64)

In [94]:
# Format for easier understanding
cm_df = pd.DataFrame(cm)
cm_df.columns = ['Predicted 0','Predicted 1']
cm_df = cm_df.rename(index={0: 'Actual 0',1:'Actual 1'})
cm_df

Predicted 0  Predicted 1
Actual 0          755            0
Actual 1            1          869

#### Our model predicted '0' correctly 755 times while NEVER predicting '0' incorrectly 
#### Also it predicted  '1'  correctly 869 times while predicting '1' incorrectly  ONCE

### OTHER MODELS

In [95]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier # for K nearest neighbours
from sklearn import svm #for Support Vector Machine (SVM) 

In [96]:
dt = DecisionTreeClassifier()
dt.fit(x_train,y_train)
y1 = dt.predict(x_test)
acc1 = metrics.accuracy_score(y1, y_test)
acc1

1.0

In [97]:
kk = KNeighborsClassifier()
kk.fit(x_train,y_train)
y2 = kk.predict(x_test)
acc2 = metrics.accuracy_score(y2, y_test)
acc2

1.0

In [98]:
sv = svm.SVC()
sv.fit(x_train,y_train)
y3 = sv.predict(x_test)
acc3 = metrics.accuracy_score(y3, y_test)
acc3

C:\Users\ABOLARIN BUKOLA\Anaconda3\lib\site-packages\sklearn\svm\base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


0.9987692307692307

#### After comparison with some other model we see that Decision tree and KNeigbors both gave us 100% accuracy but our model was close enough with ~99.9% accuracy 

###  CONCLUSION
#### Let's try to make a table and interpret what weight(BIAS) and odds means

In [99]:
pd.options.display.max_rows = 999
result = pd.DataFrame(data = x.columns, columns = ['Features'])
result['BIAS'] = np.transpose(reg.coef_)
result['odds'] = np.exp(np.transpose(reg.coef_))
result

Features      BIAS       odds
0                  cap-shape_c -0.662813   0.515399
1                  cap-shape_f  0.331583   1.393172
2                  cap-shape_k  0.145572   1.156701
3                  cap-shape_s  0.603831   1.829112
4                  cap-shape_x  0.386616   1.471991
5                cap-surface_g -0.971633   0.378465
6                cap-surface_s -0.573300   0.563662
7                cap-surface_y -0.485270   0.615531
8                  cap-color_c  1.172091   3.228737
9                  cap-color_e -0.132953   0.875507
10                 cap-color_g  0.136360   1.146095
11                 cap-color_n  0.393466   1.482108
12                 cap-color_p -0.620913   0.537454
13                 cap-color_r  0.553687   1.739656
14                 cap-color_u  0.466550   1.594484
15                 cap-color_w -0.194445   0.823291
16                 cap-color_y  0.567545   1.763932
17                   bruises_t  0.458114   1.581090
18                      odor_c -3.449652   0.031757
19                      odor_f -2.979760   0.050805
20                      odor_l  1.583401   4.871494
21                      odor_m -0.265080   0.767145
22                      odor_n  4.075377  58.872646
23                      odor_p -3.056877   0.047034
24                      odor_s -1.332044   0.263937
25                      odor_y -1.319339   0.267312
26           gill-attachment_f -0.340580   0.711358
27              gill-spacing_w  2.883705  17.880399
28                 gill-size_n -3.349832   0.035090
29                gill-color_e  0.637926   1.892552
30                gill-color_g  0.130056   1.138892
31                gill-color_h -0.029045   0.971373
32                gill-color_k  0.487861   1.628828
33                gill-color_n  0.850859   2.341658
34                gill-color_o  0.194644   1.214879
35                gill-color_p  0.685126   1.984021
36                gill-color_r -0.672471   0.510446
37                gill-color_u  0.281993   1.325769
38                gill-color_w  0.472725   1.604360
39                gill-color_y -0.118779   0.888004
40               stalk-shape_t  1.006444   2.735855
41                stalk-root_b -2.287043   0.101566
42                stalk-root_c  1.512612   4.538569
43                stalk-root_e -0.793218   0.452387
44                stalk-root_r  1.359912   3.895850
45  stalk-surface-above-ring_k -1.821912   0.161716
46  stalk-surface-above-ring_s  0.747400   2.111503
47  stalk-surface-above-ring_y -0.267929   0.764962
48  stalk-surface-below-ring_k -0.664437   0.514563
49  stalk-surface-below-ring_s -0.122966   0.884294
50  stalk-surface-below-ring_y -1.468884   0.230182
51    stalk-color-above-ring_c -0.265080   0.767145
52    stalk-color-above-ring_e  0.533006   1.704048
53    stalk-color-above-ring_g  0.341585   1.407177
54    stalk-color-above-ring_n  0.261250   1.298553
55    stalk-color-above-ring_o  0.530141   1.699172
56    stalk-color-above-ring_p -0.444549   0.641114
57    stalk-color-above-ring_w  0.093327   1.097821
58    stalk-color-above-ring_y -0.819940   0.440458
59    stalk-color-below-ring_c -0.265080   0.767145
60    stalk-color-below-ring_e  0.436463   1.547225
61    stalk-color-below-ring_g  0.285971   1.331054
62    stalk-color-below-ring_n  1.128679   3.091569
63    stalk-color-below-ring_o  0.530141   1.699172
64    stalk-color-below-ring_p -0.600115   0.548749
65    stalk-color-below-ring_w -0.286382   0.750976
66    stalk-color-below-ring_y -1.077789   0.340347
67                veil-color_o  0.228796   1.257085
68                veil-color_w  0.307247   1.359677
69                veil-color_y -0.819940   0.440458
70               ring-number_o  0.020579   1.020793
71               ring-number_t  0.261949   1.299461
72                 ring-type_f  1.839448   6.293063
73                 ring-type_l -0.639883   0.527354
74                 ring-type_n -0.265080   0.767145
75                 ring-type_p  0.612028   1.844167
76         spore-p

In [100]:
#T be able to identify our refernce model
df['cap-shape'].unique()

array(['x', 'b', 's', 'f', 'k', 'c'], dtype=object)

#### Using feature cap-shape as an example cap-shape_b is our baseline model i.e all cap-shape feature will be referenced to it and note that it is the only one that doesn't appear in the table above

#### From the dataset description which can be found on kaggle 'b' represent bell while 's' represent sunken
#### To interpret cap-shape_s in terms of odds we can say that  cap-shape_s  is almost twice more likely to cause a change in our target varable than cap-shape_b(our baseline model)

#### If you find this notebook useful don't forget to upvote. #Happycoding